core/src/main/java/edu/wpi/grip/core/http/GripServer.java

```diff
@@ -0,0 +1,404 @@
+     *
+     * @param name the name to check
+     * @return true if the name is associated with a supplier, false if not
+     */
+    public boolean hasDataSupplier(String name) {
+        return dataSuppliers.containsKey(name);
+    }
+
+    /**
+     * Removes the data supplier for data with the given name
+     *
+     * @param name the name of the data to remove
+     */
+    public void removeDataSupplier(String name) {
+        dataSuppliers.remove(name);
+    }
+
+    private String createJson(Map<String, List<String>> params) {
+        final Gson gson = new GsonBuilder()
+                .setPrettyPrinting()
+                .serializeSpecialFloatingPointValues()
+                .create();
+        final Map<String, Object> data = new HashMap<>();
+        dataSuppliers.entrySet()
+                .stream()
+                .filter(e -> params.isEmpty() || params.containsKey(e.getKey())) // send everything if nothing is specified
+                .forEach(e -> data.put(e.getKey(), e.getValue().get()));
+        try {
+            return gson.toJson(data);
+        } catch (Throwable t) {
+            logger.log(Level.SEVERE, "Error generating json response", t);
+            throw new GripException(t);
+        }
+    }
+
+    private boolean didStart = false;
+    private boolean stopped = false;
+
+    /**
+     * Starts this server. Has no effect if the server has already been started.
+     */
+    public void start() {
+        if (!didStart && !stopped) {
+            // Don't call server.start() if it's already been started or stopped
+            server.start();
+            didStart = true;
+            stopped = false;
+        }
+    }
+
+    /**
+     * Stops this server. Note that a shutdown hook has been registered to call
+     * this method, so it's unlikely that this should need to be called.
+     */
+    public void stop() {
+        if(!stopped) {
+            server.stop(0);
+            didStart = false;
+            stopped = true;
+        }
+    }
+    
+    
+    /**
+     * Restarts the server on the current port.
+     */
+    public void restart() {
+        try {
+            stop();
+            server = HttpServer.create(new InetSocketAddress("localhost", getPort()), BACKLOG);
+            start();
+        } catch (IOException | IllegalStateException ex) {
+            Logger.getLogger(GripServer.class.getName()).log(Level.SEVERE, null, ex);
+            throw new GripException("Could not restart GripServer", ex);
+        }
+    }
+    
+    /**
+     * Gets the port this server is running on.
+     */
+    public int getPort() {
+        return server.getAddress().getPort();
+    }
+
+    /**
+     * Parses request parameters from a URI. For example, the URI
+     * "http://roborio-190-frc.local:8080/GRIP/data?foo=1;bar=2" will be parsed to
+     * {"foo"="1", "bar"="2"}.
+     */
+    private static Map<String, List<String>> getRequestParameters(final URI requestUri) {
```

Comment:
What? You said earlier that we'd really only need to test `public` methods. And everything runs inside that `if` statement....